File: tests/js/types_test.js
import { assert, assertEquals } from "./testutils"

function checkArgument(env) {
  assertEquals(env.namespace, 'namespace')
  assertEquals(env.candidate.text, 'text')
  assertEquals(env.engine.schema.id, '.default')
  env.candidate.text = 'new text'

  const config = env.engine.schema.config
  assertEquals(config.getBool('key'), null)
  assertEquals(config.getBool('key1'), true)
  assertEquals(config.getBool('key2'), false)
  assertEquals(config.getInt('key3'), 666)
  assertEquals(config.getDouble('key4'), 0.999)
  assertEquals(config.getString('key5'), 'string')

  const list = config.getList('list')
  assertEquals(list.getValueAt(0).getString(), 'item1')
  assertEquals(list.getValueAt(1).getString(), 'item2')
  assertEquals(list.getValueAt(2).getString(), 'item3')
  assertEquals(list.getValueAt(3), null)

  config.setString('greet', 'hello from js')

  const context = env.engine.context
  assertEquals(context.input, 'hello')

  assert(context.preedit !== null, 'preedit should not be null')
  assertEquals(context.preedit.text, 'hello', 'preedit should have text')
  assertEquals(context.preedit.caretPos, 5, 'preedit should have caretPos')
  assertEquals(context.preedit.selectStart, 0, 'preedit should have selectStart')
  assertEquals(context.preedit.selectEnd, 0, 'preedit should have selectEnd')

  context.input = 'world'

  env.newCandidate = new Candidate('js', 32, 100, 'the text', 'the comment', 888)

  // ensure adding extra fields to the qjs object would not break the quickjs engine
  env.newCandidate.extraField = 'extra field'
  assertEquals(env.newCandidate.extraField, 'extra field')

  testEnvUtilities(env)
  testTrie(env)

  return env
}

function testEnvUtilities(env) {
  const info = env.getRimeInfo()
  console.log(`Rime info = ${info}`)
  assert(info.includes('libRime v'))
  assert(info.includes('libRime-qjs v'))
  assert(info.includes('Process RSS Mem: '))
  assert(info.includes('QuickJS Mem: '))

  console.error(`This is an error message.`)

  // ensure engine.processKey would not crash the program
  env.engine.processKey('Down')
  env.engine.processKey('InvalidKey')

  assertEquals(env.popen(`echo libRime-qjs`).trim(), 'libRime-qjs')

  assertEquals(env.fileExists(env.currentFolder + '/js/types_test.js'), true)
  assertEquals(env.fileExists(env.currentFolder + '/js/not_found.js'), false)

  // test load file with utf-8 chars
  const content = env.loadFile(env.currentFolder + '/js/types_test.js')
  assert(content.includes('测试 UTF-8 编码'))
}

function testTrie(env) {
  const trie = new Trie()
  trie.loadTextFile(env.currentFolder + '/dummy_dict.txt', 6)
  checkTrieData(trie)

  trie.saveToBinaryFile(env.currentFolder + '/dumm.bin')
  const trie2 = new Trie()
  trie2.loadBinaryFile(env.currentFolder + '/dumm.bin')
  checkTrieData(trie2)
}

function checkTrieData(trie) {
  const result1 = trie.find('accord')
  assertEquals(result1, '[ә\'kɒ:d]; n. 一致, 调和, 协定\\n vt. 给与, 使一致\\n vi. 相符合')
  const result2 = trie.find('accordion')
  assertEquals(result2, '[ә\'kɒ:djәn]; n. 手风琴\\n a. 可折叠的')
  const result3 = trie.find('nonexistent-word')
  assertEquals(result3, null)
  const prefix_results = trie.prefixSearch('accord')
  assertEquals(prefix_results.length, 6)
}


globalThis.checkArgument = checkArgument

const load_file_test_data = [
  'Hello, 世界!',
  '测试 UTF-8 编码',
  '🌟 Emoji test',
  'Mixed content: あいうえお',
]

File: tests/types.test.cpp
#include <gtest/gtest.h>
#include <rime/candidate.h>
#include <rime/config/config_component.h>
#include <rime/context.h>
#include <rime/engine.h>
#include <rime/schema.h>

#include "jsvalue_raii.h"
#include "qjs_candidate.h"
#include "qjs_engine.h"
#include "qjs_environment.h"
#include "trie_data_helper.h"

using namespace rime;

class QuickJSTypesTest : public ::testing::Test {
private:
  TrieDataHelper trieDataHelper_ = TrieDataHelper("./tests", "dummy_dict.txt");

protected:
  void SetUp() override {
    QjsHelper::basePath = "tests/js";
    trieDataHelper_.createDummyTextFile();
  }

  void TearDown() override {
    trieDataHelper_.cleanupDummyFiles();
    std::remove("tests/dumm.bin");  // the file generated in js
  }
};

// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables, readability-function-cognitive-complexity)
TEST_F(QuickJSTypesTest, WrapUnwrapRimeGears) {
  auto* ctx = QjsHelper::getInstance().getContext();

  auto* engine = Engine::Create();
  // engine->ApplySchema(&schema); // ApplySchema 会触发回调函数，导致 segfault
  // engine->schema()->schema_id() = .default, engine->schema()->schema_name() = .default
  ASSERT_TRUE(engine->schema() != nullptr);
  auto* config = engine->schema()->config();
  ASSERT_TRUE(config != nullptr);
  config->SetBool("key1", true);
  config->SetBool("key2", false);
  constexpr int A_INT_NUMBER = 666;
  config->SetInt("key3", A_INT_NUMBER);
  constexpr double A_DOUBLE_NUMBER = 0.999;
  config->SetDouble("key4", A_DOUBLE_NUMBER);
  config->SetString("key5", "string");

  auto list = New<ConfigList>();
  list->Append(New<ConfigValue>("item1"));
  list->Append(New<ConfigValue>("item2"));
  list->Append(New<ConfigValue>("item3"));
  config->SetItem("list", list);

  auto* context = engine->context();
  ASSERT_TRUE(context != nullptr);
  context->set_input("hello");

  JSValueRAII env = QjsEnvironment::create(ctx, engine, "namespace");
  auto candidate = New<SimpleCandidate>("mock", 0, 1, "text", "comment");
  JS_SetPropertyStr(ctx, env, "candidate", QjsCandidate::Wrap(ctx, candidate));

  JSValueRAII module = QjsHelper::loadJsModuleToGlobalThis(ctx, "types_test.js");

  JSValueRAII global = JS_GetGlobalObject(ctx);
  JSValueRAII jsFunc = JS_GetPropertyStr(ctx, global, "checkArgument");
  JSValueRAII retValue = JS_Call(ctx, jsFunc, JS_UNDEFINED, 1, env.getPtr());

  JSValueRAII retJsEngine = JS_GetPropertyStr(ctx, retValue, "engine");
  Engine* retEngine = QjsEngine::Unwrap(ctx, retJsEngine);
  ASSERT_EQ(retEngine, engine);
  ASSERT_EQ(retEngine->schema()->schema_name(), engine->schema()->schema_name());
  JSValueRAII retJsCandidate = JS_GetPropertyStr(ctx, retValue, "candidate");
  an<Candidate> retCandidate = QjsCandidate::Unwrap(ctx, retJsCandidate);
  ASSERT_EQ(retCandidate->text(), "new text");
  ASSERT_EQ(retCandidate.get(), candidate.get());

  string greet;
  bool success = retEngine->schema()->config()->GetString("greet", &greet);
  ASSERT_TRUE(success);
  ASSERT_EQ(greet, "hello from js");

  Context* retContext = retEngine->context();
  ASSERT_EQ(retContext->input(), "world");

  // js code: env.newCandidate = new Candidate('js', 32, 100, 'the text', 'the comment', 888)
  JSValueRAII retJsNewCandidate = JS_GetPropertyStr(ctx, retValue, "newCandidate");
  auto newCandidate = QjsCandidate::Unwrap(ctx, retJsNewCandidate);
  ASSERT_EQ(newCandidate->type(), "js");
  ASSERT_EQ(newCandidate->start(), 32);
  ASSERT_EQ(newCandidate->end(), 100);
  ASSERT_EQ(newCandidate->text(), "the text");
  ASSERT_EQ(newCandidate->comment(), "the comment");
  ASSERT_EQ(newCandidate->quality(), 888);
}
